File: apps/web/test/return-path.test.ts
// @vitest-environment node

import { describe, expect, it } from 'vitest';

import {
  requiresSession,
  signedOutRedirect,
  validateReturnPath,
} from '../app/utils/returnPath';

describe('login and registration return path', () => {
  it.each([
    ['/app/resumes', '/app/resumes'],
    [
      '/app/resumes/0190f0e2-0000-7000-8000-000000000001',
      '/app/resumes/0190f0e2-0000-7000-8000-000000000001',
    ],
    [
      '/authorize?client_id=c&redirect_uri=https%3A%2F%2Fclient.example%2Fcb'
      + '&state=a%5Cb',
      '/authorize?client_id=c&redirect_uri=https%3A%2F%2Fclient.example%2Fcb'
      + '&state=a%5Cb',
    ],
    ['/' + 'a'.repeat(2047), '/' + 'a'.repeat(2047)],
    ['/app/resumes/a..b/.well', '/app/resumes/a..b/.well'],
    ['/authorize?path=/../x', '/authorize?path=/../x'],
  ])('keeps the same-origin path %s', (value, want) => {
    expect(validateReturnPath(value)).toBe(want);
  });

  it.each([
    undefined,
    null,
    ['/app/resumes'],
    '',
    'app/resumes',
    '//evil.example',
    'https://evil.example',
    'javascript:alert(1)',
    '/\\evil.example',
    '/%2F%2Fevil.example',
    '/%2Fevil.example',
    '/%5Cevil.example',
    '/app/%5C%5Cevil',
    '/\t/evil.example',
    '/app\n/x',
    '/app\r/x',
    '/app\x00',
    '/app\x7f',
    '/%09/evil.example',
    '/app%0A',
    '/app/%zz',
    '/./x',
    '/../x',
    '/.//evil.example',
    '/%2e//evil.example',
    '/app/new/..',
    '/a/%2E%2E/b',
    '/' + 'a'.repeat(2048),
  ])('rejects %j', (value) => {
    expect(validateReturnPath(value)).toBeNull();
  });

  it.each([
    ['/app/new', '/app/new'],
    ['/app/new?sample=ats-plain&lng=vi', '/app/new?sample=ats-plain&lng=vi'],
    [
      '/app/new?template=engineer-compact',
      '/app/new?template=engineer-compact',
    ],
    [
      '/app/new?sample=ats-plain&next=//evil.example',
      '/app/new?sample=ats-plain',
    ],
    ['/app/new?sample=ats-plain&sample=minimal-air&lng=en', '/app/new?lng=en'],
    ['/app/new?sample=unknown&template=../x', '/app/new'],
    ['/app/new?lng=en-US', '/app/new'],
    ['/app/new?lng=VI&template=ats-plain', '/app/new?template=ats-plain'],
    ['/app/new?lng=en#fragment', '/app/new?lng=en'],
    ['/APP/New/?sample=ats-plain&x=1', '/app/new?sample=ats-plain'],
    ['/app/%6Eew?x=1', '/app/new'],
  ])('keeps only the checked query of %s', (value, want) => {
    expect(validateReturnPath(value)).toBe(want);
  });
});

describe('requiresSession', () => {
  it.each([
    '/app/resumes',
    '/app/new?sample=x',
    '/app/settings/sessions',
    '/APP/Settings/Sessions/',
    '/authorize?client_id=a',
    '/Authorize/',
  ])('is true for %s', (path) => {
    expect(requiresSession(path)).toBe(true);
  });

  it.each([
    '/',
    '/app',
    '/apps',
    '/application/x',
    '/templates',
    '/login',
    '/login?next=/app/resumes',
  ])('is false for %s', (path) => {
    expect(requiresSession(path)).toBe(false);
  });
});

describe('signedOutRedirect', () => {
  it('sends /app/settings/sessions to login with next', () => {
    expect(signedOutRedirect('/app/settings/sessions')).toBe(
      '/login?next=%2Fapp%2Fsettings%2Fsessions',
    );
  });

  it('sends /app/new to register, dropping unchecked query values', () => {
    const next = encodeURIComponent(
      '/app/new?sample=engineer-compact&lng=en',
    );
    expect(
      signedOutRedirect(
        '/app/new?sample=engineer-compact&lng=en&x=1',
      ),
    ).toBe(`/register?next=${next}`);
  });

  it('keeps the full query of /authorize in next', () => {
    const next = encodeURIComponent('/authorize?client_id=a&state=b');
    expect(signedOutRedirect('/authorize?client_id=a&state=b')).toBe(
      `/login?next=${next}`,
    );
  });

  it('drops next for a path over the byte limit', () => {
    expect(signedOutRedirect('/' + 'a'.repeat(2048))).toBe('/login');
  });
});

File: apps/server/internal/auth/returnpath_test.go
package auth

import (
	"strings"
	"testing"
)

func TestValidatedLoginReturnPath(t *testing.T) {
	const fallback = "/app/resumes"
	for _, test := range []struct {
		name, raw, want string
	}{
		{"absent", "", fallback},
		{"resume list", "/app/resumes", "/app/resumes"},
		{"deep link", "/app/resumes/0190f0e2-0000-7000-8000-000000000001", "/app/resumes/0190f0e2-0000-7000-8000-000000000001"},
		{"MCP authorize keeps its query", "/authorize?client_id=c&redirect_uri=https%3A%2F%2Fclient.example%2Fcb&state=a%5Cb",
			"/authorize?client_id=c&redirect_uri=https%3A%2F%2Fclient.example%2Fcb&state=a%5Cb"},
		{"network path", "//evil.example", fallback},
		{"absolute URL", "https://evil.example", fallback},
		{"javascript scheme", "javascript:alert(1)", fallback},
		{"no leading slash", "app/resumes", fallback},
		{"backslash", `/\evil.example`, fallback},
		{"encoded double slash", "/%2F%2Fevil.example", fallback},
		{"encoded slash after the first", "/%2Fevil.example", fallback},
		{"encoded backslash", "/%5Cevil.example", fallback},
		{"encoded backslash later in the path", "/app/%5C%5Cevil", fallback},
		{"tab", "/\t/evil.example", fallback},
		{"newline", "/app\n/x", fallback},
		{"carriage return", "/app\r/x", fallback},
		{"nul", "/app\x00", fallback},
		{"delete", "/app\x7f", fallback},
		{"encoded tab", "/%09/evil.example", fallback},
		{"encoded newline", "/app%0A", fallback},
		{"bad escape", "/app/%zz", fallback},
		{"dot segment", "/./x", fallback},
		{"dot-dot segment", "/../x", fallback},
		{"dot segment before a network path", "/.//evil.example", fallback},
		{"encoded dot segment before a network path", "/%2e//evil.example", fallback},
		{"trailing dot-dot segment", "/app/new/..", fallback},
		{"encoded dot-dot segment", "/a/%2E%2E/b", fallback},
		{"dots inside a segment stay", "/app/resumes/a..b/.well", "/app/resumes/a..b/.well"},
		{"dot segment in the query stays", "/authorize?path=/../x", "/authorize?path=/../x"},
		{"2048 bytes", "/" + strings.Repeat("a", 2047), "/" + strings.Repeat("a", 2047)},
		{"2049 bytes", "/" + strings.Repeat("a", 2048), fallback},

		{"new resume", "/app/new", "/app/new"},
		{"new resume with every parameter", "/app/new?sample=ats-plain&template=ats-plain&lng=vi",
			"/app/new?lng=vi&sample=ats-plain&template=ats-plain"},
		{"only en and vi are languages", "/app/new?lng=en-US", "/app/new"},
		{"upper-case language dropped", "/app/new?lng=VI&template=ats-plain", "/app/new?template=ats-plain"},
		{"unknown parameter dropped", "/app/new?sample=ats-plain&next=//evil.example", "/app/new?sample=ats-plain"},
		{"repeated parameter dropped", "/app/new?sample=a&sample=b&lng=en", "/app/new?lng=en"},
		{"malformed id dropped", "/app/new?sample=../x&template=Ats_Plain", "/app/new"},
		{"overlong id dropped", "/app/new?template=" + strings.Repeat("a", 65), "/app/new"},
		{"malformed language dropped", "/app/new?lng=english!", "/app/new"},
		{"fragment dropped", "/app/new?lng=en#x", "/app/new?lng=en"},
		{"malformed query dropped", "/app/new?sample=a;b", "/app/new"},
		{"router-equivalent path is checked", "/APP/New/?sample=a&x=1", "/app/new?sample=a"},
		{"encoded path is checked", "/app/%6Eew?x=1", "/app/new"},
		{"encoded network path in query of a new resume", "/app/new?sample=%2F%2Fevil", "/app/new"},
	} {
		t.Run(test.name, func(t *testing.T) {
			if got := validatedLoginReturnPath(test.raw); got != test.want {
				t.Fatalf("validatedLoginReturnPath(%q) = %q, want %q", test.raw, got, test.want)
			}
		})
	}
}
